README.md
# TestCafe Reporter SMTP 
###testcafe-reporter-smtp

This is a reporter for [TestCafe](http://devexpress.github.io/testcafe). It sends the output of the test as an email via SMTP.

##Purpose
Once configured the reporter builds an email (HTML and text format) and sends via an SMTP service.  This means your test results will be emailed to you in a nicely formatted email.

##Setup instructions

    $ npm install testcafe-reporter-smtp


## Configuration

Configuration is taken from two sources:

- environment variables
- a config file on disk, can be specified via TESTCAFE_SMTP_CONFIG_FILE environment variable

The aim is that you can specify the below configuration parameters either entirely in your local environment variables, or in the config file on disk, or a mixture of both.

- TESTCAFE_SMTP_CONFIG_FILE - config file on disk in the format "VARIABLE=value", defaults to `.env`.
- TESTCAFE_SMTP_SMTPHOST - hostname of your SMTP server
- TESTCAFE_SMTP_SMTPPORT - port to connect to your SMTP server
- TESTCAFE_SMTP_SMTPUSER - username for authentication to your SMTP server
- TESTCAFE_SMTP_SMTPPASS - password for authentication to your SMTP server
- TESTCAFE_SMTP_SECURE   - true | false - passed to the `nodemailer` *secure* option (see [Security and TLS](#security-and-tls) below)
- TESTCAFE_SMTP_FROM     - From address to use when sending email. If not provided, attempts to sniff using `require("os").userInfo()`
- TESTCAFE_SMTP_TO_LIST  - comma separated list of email addresses to send reports to
- TESTCAFE_SMTP_REPORTONLYFAILURES - causes no email to be sent if all tests passed (when you only want to be notified of tests failing)
- TESTCAFE_SMTP_LOGGER   - true | false - passed to the `nodemailer` *logger* option, causes verbose output on the console
- TESTCAFE_SMTP_DEBUG    - anything | undefined - config settings are printed with console.log for debugging - *WARNING:* Will expose passwords and other sensitive information from your config!

### Security and TLS

You probably want the connection to your SMTP server to be encrypted using TLS.  From the `nodemailer` documentation:

    secure – if true the connection will use TLS when connecting to server. If false (the default) then TLS is used if server supports the STARTTLS extension. In most cases set this value to true if you are connecting to port 465. For port 587 or 25 keep it false

So, if you leave out `TESTCAFE_SMTP_SECURE`, the reporter will make a plaintext connection to your server, but if it responds with STARTTLS then the connection will be upgraded to encryption using TLS before authentication is performed.

## Testing

Run the unit tests, which set a dummy SMTP hostname that causes the reporter to not actually send the email it has built.

    $ npm test

The unit test uses Mocha, and does not actually invoke a Testcafe test run or send any emails.

## Usage with TestCafe

You can run TestCafe with testcafe-reporter-smtp either via the command line, or via a test runner.

You must have an SMTP server available to test and a username on it, e.g. gmail.com with your Google credentials.

- cd into your test project.
- Edit or create the `.env` file by adding the following required variables:

    - TESTCAFE_SMTP_SMTPHOST=gmail.com
    - TESTCAFE_SMTP_SMTPUSER=*****
    - TESTCAFE_SMTP_SMTPPASS=*****
    - TESTCAFE_SMTP_TO_LIST=your_user@example.com

Alternatively, call the config file anything you want and set `TESTCAFE_SMTP_CONFIG_FILE=<filename>`.

### Running tests from the commmand line:

```
$ testcafe chrome 'path/to/test/file.js' --reporter smtp
```

**Please Note:** The command line `testcafe` binary executes a forced process stop once the test has finished running, which will probably terminate before the SMTP communication has completed.  You may have better results using the runner API below.

### Running tests from a runner:

When you use TestCafe API, you can pass the reporter name to the `reporter()` method:

```js
const createTestCafe = require('testcafe');
let runner           = null;
let testcafe         = null;

createTestCafe('localhost', 1337, 1338)
    .then(tc => {
		testcafe = tc;
        runner = tc.createRunner();

        return runner
			.src('index.ts') // This is the test fixture to run
			.browsers(['chrome'])
			.reporter('smtp') // This tells it to look for testcafe-reporter-smtp
			.run()
			.then(failedCount => {
				console.log('Runner finished, failedCount: ', failedCount);
				testcafe.close();
			})
			.catch(error => {
				console.error('Caught error: ', error);
			});
});
```

##Further Documentation
[TestCafe Reporter Plugins](https://devexpress.github.io/testcafe/documentation/extending-testcafe/reporter-plugin/)


changelog.md
# Changelog
All notable changes to this project will be documented in this file.

## [Unreleased]

## [0.1.0] - 2021-06-29
### Added
- TESTCAFE_SMTP_DEBUG config variable to perform some console.log statements.
- Actual tests that compare output against stored expected output.

### Changed
- Ported into a fresh project built using the now-recommended Yeoman approach.  This resulted in updated versions of Gulp and various other tools, removing some old security vulnerabilities.

## [0.0.2] - some time in 2018
### Added
- Built by copying a Testcafe reporter and replacing key parts.

Gulpfile.js
var gulp    = require('gulp');
var eslint  = require('gulp-eslint');
var babel   = require('gulp-babel');
var mocha   = require('gulp-mocha');
var del     = require('del');

function clean (cb) {
    del('lib', cb);
}

function lint () {
    return gulp
        .src([
            'src/**/*.js',
            'test/**/*.js',
            'Gulpfile.js'
        ])
        .pipe(eslint())
        .pipe(eslint.format())
        .pipe(eslint.failAfterError());
}

function copyTemplates () {
    return gulp
        .src('src/templates/*')
        .pipe(gulp.dest('lib/templates'));
}

function build () {
    return gulp
        .src('src/**/*.js')
        .pipe(babel())
        .pipe(gulp.dest('lib'));
}

function test () {
    process.env.TESTCAFE_SMTP_SMTPHOST = 'unit-test-mock';
    return gulp
        .src('test/test.js')
        .pipe(mocha({
            ui:       'bdd',
            reporter: 'spec',
            timeout:  typeof v8debug === 'undefined' ? 20000 : Infinity // NOTE: disable timeouts in debug
        }));
}

function preview () {
    var buildReporterPlugin = require('testcafe').embeddingUtils.buildReporterPlugin;
    var pluginFactory       = require('./lib');
    var reporterTestCalls   = require('./test/utils/reporter-test-calls');
    var plugin              = buildReporterPlugin(pluginFactory);

    console.log();

    reporterTestCalls.forEach(function (call) {
        plugin[call.method].apply(plugin, call.args);
    });

    process.exit(0);
}

exports.clean = clean;
exports.lint = lint;
exports.test = gulp.series(clean, lint, build, copyTemplates, test);
exports.build = gulp.series(clean, lint, build, copyTemplates);
exports.preview = gulp.series(clean, lint, build, copyTemplates, preview);


src/MailMessage.js
const nodemailer = require('nodemailer');
const handlebars = require('handlebars');
const fs = require('fs');

function MailMessage (info, config) {
    this.info = info;
    this.config = config;
}

MailMessage.prototype.send = function send () {
    // First compile the template from handlebars source.
    const htmlTemplate = fs.readFileSync(this.config.htmlTemplateFile, 'utf-8');
    const htmlTmpl = handlebars.compile(htmlTemplate);
    const textTemplate = fs.readFileSync(this.config.textTemplateFile, 'utf-8');
    const textTmpl = handlebars.compile(textTemplate);

    // Run our data object through the template to produce the email body.
    const htmlBody = htmlTmpl(this.info);
    const textBody = textTmpl(this.info);
    // console.log('Produced output: ' + htmlBody);

    // Prepare email transport
    if (this.config.bDebug) console.log('Transport options: ', this.config.smtpOptions);
    const transporter = nodemailer.createTransport(this.config.smtpOptions);

    // setup email data with unicode symbols
    const mailOptions = {
        from:    this.info.from, // sender address
        to:      this.info.recipients,
        subject: this.info.subject, // Subject line
        text:    textBody, // plain text body
        html:    htmlBody // html body
    };

    // console.log('mailOptions: ', mailOptions);
    // console.log('smtpOptions: ', this.config.smtpOptions);

    if (this.config.smtpOptions.host === 'unit-test-mock') {
        // Gulpfile.js has indicated we are running a unit test, don't try to actually send an email.
        console.log('Unit Test Mock Host defined - do not send emails but consider run successful.');
        console.log('Text email body was:\n********************\n' + textBody + '\n******************');
    }
    else {
        // send mail with defined transport object
        transporter.sendMail(mailOptions, (error, info) => {
            if (error) {
                console.log(error);
                return;
            }
            console.log('Message sent: %s', info.messageId);

            // Message sent: <b658f8ca-6296-ccf4-8306-87d57a0b4321@example.com>
            // Preview URL: https://ethereal.email/message/WaQKMgKddxQDoou...
        });
    }

    return textBody;
};

module.exports = MailMessage;

// module.exports = function () {
//     return {
//         /**
//          * info - object containing email contents information, to be passed into the Handlebars template.
//          * config - configuration e.g. smtpOptions.
//          */
//         constructor (info, config) {
//             console.log(`MailMessage ctor here, config ${config}`);
//             this.info = info;
//             this.config = config;
//         },

//         send () {
//             // First compile the template from handlebars source.
//             const htmlTemplate = fs.readFileSync(this.config.htmlTemplateFile, 'utf-8');
//             const htmlTmpl = handlebars.compile(htmlTemplate);
//             const textTemplate = fs.readFileSync(this.config.textTemplateFile, 'utf-8');
//             const textTmpl = handlebars.compile(textTemplate);

//             // Run our data object through the template to produce the email body.
//             const htmlBody = htmlTmpl(this.info);
//             const textBody = textTmpl(this.info);
//             // console.log('Produced output: ' + htmlBody);

//             // Prepare email transport
//             // console.log('Transport options: ', this.config.smtpOptions);
//             const transporter = nodemailer.createTransport(this.config.smtpOptions);

//             // setup email data with unicode symbols
//             // console.log('User info: ', userInfo);
//             const mailOptions = {
//                 from:    this.info.from, // sender address
//                 to:      this.info.recipients,
//                 subject: this.info.subject, // Subject line
//                 text:    textBody, // plain text body
//                 html:    htmlBody // html body
//             };

//             // console.log('mailOptions: ', mailOptions);
//             // console.log('smtpOptions: ', this.config.smtpOptions);

//             if (this.config.smtpOptions.host === 'unit-test-mock') {
//                 // Gulpfile.js has indicated we are running a unit test, don't try to actually send an email.
//                 console.log('Unit Test Mock Host defined - do not send emails but consider run successful.');
//                 console.log('Text email body was:\n********************\n' + textBody + '\n******************');
//             }
//             else {
//                 // send mail with defined transport object
//                 transporter.sendMail(mailOptions, (error, info) => {
//                     if (error) {
//                         console.log(error);
//                         return;
//                     }
//                     console.log('Message sent: %s', info.messageId);

//                     // Message sent: <b658f8ca-6296-ccf4-8306-87d57a0b4321@example.com>
//                     // Preview URL: https://ethereal.email/message/WaQKMgKddxQDoou...
//                 });
//             }
//         }
//     };
// };


src/index.js
/**
 * Read config from .env file or the file specified by TESTCAFE_SMTP_CONFIG_FILE.
 * Example config file:
 *   TESTCAFE_SMTP_SMTPHOST=example.com
 *   TESTCAFE_SMTP_SMTPUSER=mickey
 *   TESTCAFE_SMTP_SMTPPASS=mouse
 *   TESTCAFE_SMTP_SMTPPORT=25
 *   TESTCAFE_SMTP_TO_LIST=me@example.com,mum@example.com
 * Email construction and sending is performed in MailMessage, and is separated
 * from the config mechanism (all values are passed in as objects).
 * @type Module envs|Module envs
 */
const MailMessage = require('./MailMessage');
const dotenv = require('dotenv');
const configSource = process.env.TESTCAFE_SMTP_CONFIG_FILE;
const bDebug = process.env.TESTCAFE_SMTP_DEBUG;
const dotenvOptions = configSource ? { path: configSource } : null;
const path = require('path');

dotenv.config(dotenvOptions);

function getEnv (key, _default) {
    return process.env.hasOwnProperty(key) ? process.env[key] : _default;
}

module.exports = function () {
    return {
        noColors: true,

        reportTaskStart (startTime, userAgents, testCount) {
            // Include secure if it is defined, else let it default.
            if (bDebug) console.log(`Loading config from ${configSource}... dotenv options ${JSON.stringify(dotenvOptions)}`);
            var secure = getEnv('TESTCAFE_SMTP_SECURE', '');

            this.reportOnlyFailures = getEnv('TESTCAFE_SMTP_REPORTONLYFAILURES', 'true');
            this.startTime = startTime;
            this.testCount = testCount;
            this.fixtureInfo = {}; // keyed by fixture name, data for email template
            this.messages = [];
            this.smtpOptions = {
                host: getEnv('TESTCAFE_SMTP_SMTPHOST', 'smtp.example.com'),
                port: getEnv('TESTCAFE_SMTP_SMTPPORT', 587),
                auth: {
                    user: getEnv('TESTCAFE_SMTP_SMTPUSER', 'username'),
                    pass: getEnv('TESTCAFE_SMTP_SMTPPASS', 'password')
                },
                logger: getEnv('TESTCAFE_SMTP_LOGGER', 'false') === 'true'
            };
            if (secure)
                this.smtpOptions.secure = secure;
            if (bDebug) console.log(`Built smtpOptions: ${JSON.stringify(this.smtpOptions)}`);

            this.messages.push(`Starting testcafe ${startTime}.\n Running tests in: ${userAgents}`);
        },

        reportFixtureStart (name/*, path*/) {
            this.currentFixtureName = name;
            this.fixtureInfo[name] = {
                tests: {}
            };
        },

        reportTestStart (/* name, testMeta */) {
            // NOTE: This method is optional.
        },

        reportTestDone (name, testRunInfo) {
            // Pretty format errors to a new property
            if (testRunInfo.errs.length > 0) {
                testRunInfo.formattedErrs = testRunInfo.errs.map((error, id) => {
                    return this.formatError(error, `${id + 1} `);
                });
            }
            // console.log('Test (' + name + ') results: ', testRunInfo);
            // Record this test's results in the hash for this fixture.
            this.fixtureInfo[this.currentFixtureName].tests[name] = testRunInfo;
        },

        reportTaskDone (endTime, passed, warnings) {
            // console.log(`endTime ${endTime}, warnings ${warnings}`);
            // const durationMs  = endTime - this.startTime;
            const failedCount = this.testCount - passed;

            let from = getEnv('TESTCAFE_SMTP_FROM', null);

            const subject = failedCount === 0 ?
                `Tests passed OK : ${this.testCount} passed of ${this.testCount}` :
                `TESTS FAILING : ${failedCount} of ${this.testCount} failed`;

            if (failedCount === 0 && getEnv('TESTCAFE_SMTP_REPORTONLYFAILURES', '') === 'true') {
                console.log('reporter-smtp: No failures, not sending email.');
                return;
            }

            if (!from) {
                const userInfo = require('os').userInfo();

                from = `"${userInfo.username}" <${userInfo.email}>`;
            }

            // Pass all our data to the email builder for templating and sending.
            const mail = new MailMessage({
                from:       from,
                recipients: getEnv('TESTCAFE_SMTP_TO_LIST', 'test@example.com, test2@example.com'), // list of receivers
                subject,
                passed,
                failedCount,
                warnings,
                messages:   this.messages,
                fixtures:   this.fixtureInfo
            }, {
                bDebug:           bDebug,
                smtpOptions:      this.smtpOptions,
                htmlTemplateFile: getEnv('TESTCAFE_SMTP_HTML_EMAIL_TEMPLATE', path.join(__dirname, 'templates/defaultHTML.handlebars')),
                textTemplateFile: getEnv('TESTCAFE_SMTP_TEXT_EMAIL_TEMPLATE', path.join(__dirname, 'templates/defaultTEXT.handlebars'))
            });

            const textBody = mail.send();

            this.write(textBody);
        }
    };
};
